File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.Stack;
class Solution {
    boolean solution(String s) {
        Stack<Character> st = new Stack<>();
        
        for (char c : s.toCharArray())
        {
            if (c == '(')
                st.push(c);
            else if (c == ')') {
                if (st.empty() || '(' != st.peek())
                    return false;
                st.pop();
            }
        }        
        return st.empty();
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include <string>
#include <iostream>
#include <stack>
using namespace std;

bool solution(string s)
{
    int i = 0;
    stack<char> st;
    while (s[i]) {
        if (s[i] == '(')
            st.push(s[i]);
        else if (s[i] == ')') {
            if (st.empty() || st.top() != '(')
                return false;
            else
                st.pop();
        }
        else
            throw new exception;
        i++;
    }
    return st.empty();
}